### PR TITLE
[FIRRTL][LowerToHW] Generate concurrent verification statements

### PR DESCRIPTION
**Pull request by prithayan**
Add support for lowering to concurrent instead of immediate verification statements.
This emulates the SFC behavior of lowering printf encoded verification statements to concurrent statements.
1. Added `isConcurrent` attribute to `FIRRTL`  `assert/assume/cover`, and set the attribute only when lowering printf encoded statements.
2. Added a enum attribute `eventControl` with default value set to `AtPosEdge`, this can be used to represent the `sv::EventControlAttr`.  FIRRTL dialect is not using the `eventControl` attribute currently.
3. Also update the `LowerToHW` to lower the corresponding concurrent statements, depending on the `isConcurrent` attribute.
4. Add corresponding test cases.

### CODE REVIEW

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 1]
```diff
@@ -2759,58 +2791,102 @@ LogicalResult FIRRTLLowering::visitStmt(StopOp op) {
...
+    event = circt::sv::EventControl::AtPosEdge;
+  else if (op.eventControl() == "AtEdge")
+    event = circt::sv::EventControl::AtEdge;
+  else if (op.eventControl() == "AtNegEdge")
+    event = circt::sv::EventControl::AtNegEdge;
+  else {
+    op.emitError(
+        "Event Control can only be one of :'AtPosEdge/AtEdge/AtNegEdge'");
+    return failure();
+  }
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToHW/LowerToHW.cpp:2847]: I'm not big on using a string here, but if we have to, this begs for a `StringSwitch` that assigns the `event` directly.

Also, if this was an enumeration of some kind, then we wouldn't need to do verification of the attribute during lowering. 🤔

**Review by seldridge — COMMENTED** [review]
I like the solution of the boolean storing concurrent or not in FIRRTL Dialect. 👍 

I'm skeptical of using a string for the event control, though.  Is there a reason why we can't use the SV attribute here?

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp (hunk 1 above)
Comment (lattner) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:2847]: Right, please move to using an enum for this.  There are a bunch of examples in ODS, e.g. `RUWAttr` in FIRRTL dialect

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 2]
```diff
@@ -2759,58 +2791,102 @@ LogicalResult FIRRTLLowering::visitStmt(StopOp op) {
...
 }
 
 // Lower a cover to SystemVerilog.
 LogicalResult FIRRTLLowering::visitStmt(CoverOp op) {
-  return lowerVerificationStatement<CoverOp, sv::CoverOp>(op, coverAnnoClass);
+  return op.isConcurrent()
+             ? lowerVerificationConcurrentStatement<CoverOp,
+                                                    sv::CoverConcurrentOp>(
+                   op, coverAnnoClass)
+             : lowerVerificationStatement<CoverOp, sv::CoverOp>(op,
```
Comment (lattner) [inline lib/Conversion/FIRRTLToHW/LowerToHW.cpp:2888]: Please merge lowerVerificationConcurrentStatement and lowerVerificationStatement to reduce the redundancy in each of the visit operations.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 3]
```diff
@@ -2323,26 +2323,26 @@ ParseResult FIRStmtParser::parsePrintf() {
   StringRef formatStringRef(formatStrUnescaped);
   if (formatStringRef.startswith("cover:")) {
     APInt constOne(1, 1, false);
-    auto constTrue =
+    Value constTrue =
         builder.create<ConstantOp>(UIntType::get(getContext(), 1), constOne);
-    builder.create<CoverOp>(clock, condition, constTrue, formatStrUnescaped,
-                            "");
+    builder.create<CoverOp>(clock, condition, constTrue, formatStrUnescaped, "",
+                            true);
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:2329]: Plz put a /*Foo=*/ comment before the true so we know what is being set. (also below)

**Review by lattner — APPROVED** [review]
Looks reasonable to me (with a couple noted changes), thanks!

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp (hunk 1 above)
Comment (prithayan) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp]: Thanks, added a FIRRTL enum attribute to fix this.

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp (hunk 2 above)
Comment (prithayan) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp]: Done, merged the functions by adding a third template argument.

**Review by lattner — APPROVED** [review]
nice, thank you!

**Pull request merged**